tests/core/test_datamodules.py

```diff
@@ -221,13 +276,14 @@ def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
     checkpoint_path = list(trainer.checkpoint_callback.best_k_models.keys())[0]
     checkpoint = torch.load(checkpoint_path)
     assert dm.__class__.__name__ in checkpoint
-    assert checkpoint[dm.__class__.__name__] == dm.__class__.__name__
+    assert checkpoint[dm.__class__.__name__] == {"my": "state_dict"}
+    assert "my_dm_state" in checkpoint
+    assert checkpoint["my_dm_state"] == {"my": "state_dict"}
 
     for trainer_fn in TrainerFn:
         trainer.state.fn = trainer_fn
-        with mock.patch.object(dm, "on_load_checkpoint") as dm_mock:
-            trainer._restore_modules_and_callbacks(checkpoint_path)
-            dm_mock.assert_called_once()
+        trainer._restore_modules_and_callbacks(checkpoint_path)
+        assert dm.my_state_dict == {"my": "state_dict", "on_load": "update"}
```

Comment:
example failure - because we introduce `load_state_dict` call after `on_load_checkpoint` - if a user does already have a `Stateful` datamodule (and does not expect lightning to call `load_state_dict`), changes made in `on_load_checkpoint` could be overridden by lightning's `load_state_dict` call after it.

What's the smoothest way to resolve/introduce this change?

Reply:
taken a step further, the user may have defined a load_state_dict/state_dict with different arguments, and the lightning trainer calling this function will result in an error because of a mismatched signature. 

I don't think there's any smooth way to get around this because it's a risk of adding to the superclass and relying on inheritance/OOP

however i think the chance of this happening is very, very small.